File: app/src/main/java/com/xianzhitech/ptt/util/ContactComparator.java
package com.xianzhitech.ptt.util;

import com.xianzhitech.ptt.model.ContactItem;
import com.xianzhitech.ptt.model.Group;

import java.text.Collator;
import java.util.Comparator;
import java.util.Locale;

/**
 * Created by fanchao on 15/12/15.
 */
public class ContactComparator implements Comparator<ContactItem> {
    private final Collator collator;

    public ContactComparator() {
        collator = Collator.getInstance();
    }

    public ContactComparator(final Locale locale) {
        collator = Collator.getInstance(locale);
    }

    @Override
    public int compare(final ContactItem lhs, final ContactItem rhs) {
        final boolean lhsIsGroup = lhs instanceof Group, rhsIsGroup = rhs instanceof Group;
        if (lhsIsGroup && rhsIsGroup || (!lhsIsGroup && !rhsIsGroup)) {
            return collator.compare(lhs.getName(), rhs.getName());
        }
        else if (lhsIsGroup) {
            return -1;
        }
        else { // if rhsIsGroups
            return 1;
        }
    }
}

File: app/src/main/kotlin/com/xianzhitech/ptt/util/ContactComparator.kt
package com.xianzhitech.ptt.util

import com.xianzhitech.ptt.model.Group
import com.xianzhitech.ptt.model.User
import java.text.Collator
import java.util.*

/**
 * Created by fanchao on 15/12/15.
 */
class ContactComparator : Comparator<Any> {
    private val collator: Collator

    constructor() {
        collator = Collator.getInstance()
    }

    constructor(locale: Locale) {
        collator = Collator.getInstance(locale)
    }

    override fun compare(lhs: Any, rhs: Any): Int {
        val lhsIsGroup = lhs is Group
        val rhsIsGroup = rhs is Group
        val lhsName = if (lhsIsGroup) (lhs as Group).name else (lhs as User).name
        val rhsName = if (rhsIsGroup) (rhs as Group).name else (rhs as User).name
        if (lhsIsGroup && rhsIsGroup || (!lhsIsGroup && !rhsIsGroup)) {
            return collator.compare(lhsName, rhsName)
        } else if (lhsIsGroup) {
            return -1
        } else {
            // if rhsIsGroups
            return 1
        }
    }
}
